src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultRazorProjectChangePublisher.cs

```diff
@@ -0,0 +1,202 @@
+                return;
+            }
+
+            _pendingProjectPublishes.Remove(projectFilePath);
+
+            Publish(projectSnapshot);
+        }
+
+        internal void Publish(ProjectSnapshot projectSnapshot)
+        {
+            if (projectSnapshot is null)
+            {
+                throw new ArgumentNullException(nameof(projectSnapshot));
+            }
+
+            lock (_publishLock)
+            {
+                string publishFilePath = null;
+                try
+                {
+                    if (!_publishFilePathMappings.TryGetValue(projectSnapshot.FilePath, out publishFilePath))
+                    {
+                        return;
+                    }
+
+                    SerializeToFile(projectSnapshot, publishFilePath);
+                }
+                catch (Exception ex)
+                {
+                    _logger.LogWarning($@"Could not update Razor project configuration file '{publishFilePath}':
+{ex}");
+                    // TODO should log
+                }
+            }
+        }
+
+        internal void RemovePublishing(ProjectSnapshot projectSnapshot)
+        {
+            lock (_publishLock)
+            {
+                var oldProjectFilePath = projectSnapshot.FilePath;
+                if (_publishFilePathMappings.TryGetValue(oldProjectFilePath, out var publishFilePath))
+                {
+                    if (_pendingProjectPublishes.TryGetValue(oldProjectFilePath, out _))
+                    {
+                        // Project was removed while a delayed publish was in flight. Clear the in-flight publish so it noops.
+                        _pendingProjectPublishes.Remove(oldProjectFilePath);
+                    }
+
+                    DeleteFile(publishFilePath);
+                }
+            }
+        }
+
+        // Virtual for testing
+        protected virtual void DeleteFile(string publishFilePath)
+        {
+            var info = new FileInfo(publishFilePath);
+            if (info.Exists)
+            {
+                try
+                {
+                    // Try catch around the delete in case it was deleted between the Exists and this delete call. This also
+                    // protects against unauthorized access issues.
+                    info.Delete();
+                }
+                catch (Exception ex)
+                {
+                    _logger.LogWarning($@"Failed to delete Razor configuration file '{publishFilePath}':
+{ex}");
+                }
+            }
+        }
+
+        protected virtual void SerializeToFile(ProjectSnapshot projectSnapshot, string publishFilePath)
+        {
+            var fileInfo = new FileInfo(publishFilePath);
+            using var writer = fileInfo.CreateText();
+            _serializer.Serialize(writer, projectSnapshot);
+        }
+
+        internal override bool IsLSPEditorAvailable(string projectFilePath)
```

Comment:
Could you clarify the purpose of having this on this type + why you can't have an importing constructor that takes in `LSPEditorFeatureDetector`?